crates/sparrow-batch/src/batch.rs

```diff
@@ -77,6 +78,35 @@ impl Batch {
         })
     }
 
+    /// Updates the up_to_time.
+    pub fn with_up_to_time(self, up_to_time: RowTime) -> Self {
+        Batch {
+            data: self.data,
+            up_to_time,
+        }
+    }
+
+    /// Replaces the batch's data with the given data.
+    ///
+    /// Panics if the [BatchInfo] is empty.
+    pub fn with_data(self, data: ArrayRef) -> Self {
+        if let Some(info) = self.data {
+            Self {
+                data: Some(BatchInfo {
+                    data,
+                    time: info.time,
+                    subsort: info.subsort,
+                    key_hash: info.key_hash,
+                    min_present_time: info.min_present_time,
+                    max_present_time: info.max_present_time,
+                }),
+                up_to_time: self.up_to_time,
+            }
+        } else {
+            panic!("Cannot replace data of empty batch");
```

Comment:
Why does this panic? Couldn't this do `self.data = Some(BatchInfo { data ... })`? Or is this because it reuses the original `time`, `subsort`, `key_hash`, etc.?

I suspect the latter, which suggests:

1. We should document this changes the data, but does not update the time, subsort, or key hash.
2. Add an assertion that the `data.len()` is the same (since otherwise we don't satisfy the requirement that the data and time/subsort/key_hash have the same number of rows).

Reply:
Yes, the latter. Updating docs accordingly, and using `BatchInfo::new()` for validations